web/src/message_scroll.js

```diff
@@ -154,12 +154,21 @@ export function initialize() {
 
         if (event.mark_read && event.previously_selected_id !== -1) {
             // Mark messages between old pointer and new pointer as read
-            let messages;
             if (event.id < event.previously_selected_id) {
-                messages = event.msg_list.message_range(event.id, event.previously_selected_id);
+                // We don't mark messages as read when the pointer moves up.
+                return;
+            }
+
+            let messages = event.msg_list.message_range(event.previously_selected_id, event.id);
+            // If the user just arrived at the message `event.id`, we don't mark it as read.
+            // We only mark messages as read when the pointer moves past the message.
+            // This is likely the last message in the list.
+            if (messages.length > 0 && messages.at(-1).id === event.id) {
+                messages.pop();
             } else {
-                messages = event.msg_list.message_range(event.previously_selected_id, event.id);
+                messages = messages.filter((msg) => msg.id !== event.id);
             }
```

Comment:
Hmm, why is this logic not just doing `event.id - 0.000001` as the end of the range? I think we need to take some care with local echo, so subtracting 1 might not be correct, but I'm not entirely sure what this logic is doing.

Also I don't know why you have this if/else statement; it seems like in either case you're removing `msg.id` from the list?

Reply:
Yeah, I thought that using something like `event.id - 0.000001` would be an easy change here but looks like `message_range` function is fairly complicated that it is trying to find the id where the message could be inserted. I didn't want to disturb that logic. So, I opted for just removing the message where the pointer is at from the range.


I am trying to optimize removing that message using the `if/else` statement. Since the `message_range` will likely have the `event.id` as the last message, using `messsage.pop` can be easy way to exclude the message, which otherwise would be a loop to find the message which is always at the last.

Changed it to just loop in the reverse order.